extendr-api/src/robj_ndarray.rs

```diff
@@ -19,31 +19,106 @@ where
+                } else {
+                    Err($error_fn(robj.clone()))
+                }
+            }
+        }
+    };
+
+    ($type: ty, $error_fn: expr) => {
+        impl<'a> TryFrom<Robj> for ArrayView1<'a, $type> {
+            type Error = crate::Error;
+
+            fn try_from(robj: &Robj) -> Result<Self> {
+                if let Some(v) = robj.as_typed_slice() {
+                    Ok(ArrayView1::<'a, $type>::from(v))
+                } else {
+                    Err($error_fn(robj.clone()))
+                }
+            }
+        }
+    };
+}
+
 macro_rules! make_array_view_2 {
-    ($type: ty, $error_str: expr) => {
+    ($type: ty, $error_str: expr, $error_fn: expr) => {
         impl<'a> FromRobj<'a> for ArrayView2<'a, $type> {
             /// Convert an R object to a `ndarray` ArrayView2.
             fn from_robj(robj: &'a Robj) -> std::result::Result<Self, &'static str> {
+                <ArrayView2<'a, $type>>::try_from(robj).map_err(|_| $error_str)
+            }
+        }
+
+        impl<'a> TryFrom<&'a Robj> for ArrayView2<'a, $type> {
+            type Error = crate::Error;
+            fn try_from(robj: &Robj) -> Result<Self> {
+                if robj.is_matrix() {
+                    let nrows = robj.nrows();
+                    let ncols = robj.ncols();
+                    if let Some(v) = robj.as_typed_slice() {
+                        // use fortran order.
+                        let shape = (nrows, ncols).into_shape().f();
+                        return ArrayView2::from_shape(shape, v)
+                            .map_err(|err| Error::NDArrayError(err));
+                    } else {
+                        return Err($error_fn(robj.clone()));
+                    }
+                }
+                return Err(Error::ExpectedMatrix(robj.clone()));
+            }
+        }
+
+        impl<'a> TryFrom<Robj> for ArrayView2<'a, $type> {
+            type Error = crate::Error;
+            fn try_from(robj: Robj) -> Result<Self> {
                 if robj.is_matrix() {
                     let nrows = robj.nrows();
                     let ncols = robj.ncols();
                     if let Some(v) = robj.as_typed_slice() {
                         // use fortran order.
                         let shape = (nrows, ncols).into_shape().f();
-                        if let Ok(res) = ArrayView2::from_shape(shape, v) {
-                            return Ok(res);
-                        }
+                        return ArrayView2::from_shape(shape, v)
+                            .map_err(|err| Error::NDArrayError(err));
+                    } else {
+                        return Err($error_fn(robj.clone()));
                     }
                 }
-                return Err($error_str);
+                return Err(Error::ExpectedMatrix(robj.clone()));
             }
         }
     };
 }
+make_array_view_1!(Rbool, Error::ExpectedLogical);
```

Comment:
Could we leverage `ToVectorValue` here to make this a generic impl instead of using macros? ie `impl<'a, T> TryFrom<Robj> for ArrayView1<'a, T> where T: ToVectorValue`.

Reply:
I also note we don't have non-R types like `i64` and `f32`. This suggestion would solve that.

Reply:
Ok so the idea behind this is that we have type-specific errors. I see no way of achieving the same using blanket implementation for all `T : ToVectorValue`. 
The case of `i64` and `f32` are different. `ArrayView` here is a view, zero-copy, and since R supports only 32bit integers and 64bit doubles, there are simply no other valid variants apart from `u32`.

Reply:
I allowed reinterpreting `integers` as `u32` and added support here.